core/embed/rust/src/ui/layout/util.rs

```diff
@@ -32,3 +40,177 @@ where
+            Ok(StrOrBytes::Bytes(obj))
+        } else {
+            Err(Error::TypeError)
+        }
+    }
+}
+
+pub struct ConfirmBlob {
+    pub description: StrBuffer,
+    pub extra: StrBuffer,
+    pub data: StrOrBytes,
+    pub description_font: &'static TextStyle,
+    pub extra_font: &'static TextStyle,
+    pub data_font: &'static TextStyle,
+}
+
+impl ParagraphSource for ConfirmBlob {
+    fn at<'a>(&'a self, index: usize, offset: usize, buffer: &'a mut [u8]) -> Paragraph<&'a str> {
+        match index {
+            0 => Paragraph::new(self.description_font, &self.description.as_ref()[offset..]),
+            1 => Paragraph::new(self.extra_font, &self.extra.as_ref()[offset..]),
+            2 => Paragraph::new(self.data_font, self.data.as_str_offset(offset, buffer)),
+            _ => unreachable!(),
+        }
+    }
+
+    fn size(&self) -> usize {
+        3
+    }
+}
+
+pub struct PropsList {
+    items: Gc<List>,
+    key_font: &'static TextStyle,
+    value_font: &'static TextStyle,
+    value_mono_font: &'static TextStyle,
+}
+
+impl PropsList {
+    pub fn new(
+        obj: Obj,
+        key_font: &'static TextStyle,
+        value_font: &'static TextStyle,
+        value_mono_font: &'static TextStyle,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            items: obj.try_into()?,
+            key_font,
+            value_font,
+            value_mono_font,
+        })
+    }
+}
+
+impl ParagraphSource for PropsList {
+    fn at<'a>(&'a self, index: usize, offset: usize, buffer: &'a mut [u8]) -> Paragraph<&'a str> {
+        let block = move |buffer| {
+            let entry = self.items.get(index / 2)?;
+            let [key, value, value_is_mono]: [Obj; 3] = iter_into_objs(entry)?;
+            let value_is_mono: bool = bool::try_from(value_is_mono)?;
+            let obj: Obj;
+            let style: &TextStyle;
+
+            if index % 2 == 0 {
+                if !key.is_str() && key != Obj::const_none() {
+                    return Err(Error::TypeError);
+                }
+                style = self.key_font;
+                obj = key;
+            } else {
+                if value_is_mono {
+                    style = self.value_mono_font;
+                } else {
+                    if value.is_type_bytes() {
+                        return Err(Error::TypeError);
+                    }
+                    style = self.value_font;
+                }
+                obj = value;
+            };
+
+            if obj == Obj::const_none() {
+                return Ok(Paragraph::new(style, ""));
+            }
+
+            let para = if obj.is_str() {
+                // SAFETY:
+                // As long as self is visible to GC, the string will be also. The paragraph
+                // rendering does not keep the returned references for long so we can reasonably
+                // expect for the chain of references from self to the buffer not to be broken.
```

Comment:
Ooooh this one is fun because the "reasonable expectation" hinges on uPy not dropping the item out of our list, correct?

that's not great.
Let's keep it as is for this PR but we'll need to implement the fully correct solution (that is StrBuffer slicing) before release

Reply:
#2665